component/attr_cache/attr_cache.go

```diff
@@ -455,34 +410,27 @@ func (ac *AttrCache) ReadDir(options internal.ReadDirOptions) (pathList []*inter
 
 // merge results from our cache into pathMap
 func (ac *AttrCache) addDirsNotInCloudToListing(listPath string, pathList []*internal.ObjAttr) ([]*internal.ObjAttr, int) {
-	prefix := dirToPrefix(listPath)
-	if prefix == "/" {
-		prefix = ""
-	}
 	numAdded := 0
-	ac.cacheLock.RLock()
-	for key, value := range ac.cacheMap {
-		// ignore invalid and deleted items
-		if !value.valid() || !value.exists() {
-			continue
-		}
-		// the only entries missing are the directories with no objects
-		if !value.attr.IsDir() || value.isInCloud() {
-			continue
-		}
-		// look for matches
-		if strings.HasPrefix(key, prefix) {
-			// exclude entries in subdirectories
-			pathInsideDirectory := strings.TrimPrefix(key, prefix)
-			if strings.Contains(pathInsideDirectory, "/") {
-				continue
-			}
 
-			pathList = append(pathList, value.attr)
-			numAdded++
+	nonCloudItem, err := ac.cacheMap.get(listPath)
+
+	if err != nil {
+		log.Err("could not find the attr cached item: ", err)
+		return nil, 0
+	}
+
+	if nonCloudItem.exists() {
+		ac.cacheLock.RLock()
+		for _, chldNonCloudItem := range nonCloudItem.children {
+			if chldNonCloudItem.exists() && !chldNonCloudItem.isInCloud() {
+				pathList = append(pathList, chldNonCloudItem.attr)
+				numAdded++
+				ac.addDirsNotInCloudToListing(chldNonCloudItem.attr.Path, pathList)
```

Comment:
This is not supposed to be recursive. ReadDir only returns the immediate contents of the directory. So this line needs to be dropped.

Reply:
didn't the original flat map design include anything that had the prefix to be counted towards this? Was that an undesirable side effect of that design? can you explain why we don't want the sub tree scanned? is that because readDir is only called for one level at a time and the end user doesn't see the whole subtree with a file system UI?